crates/fadroma-derive-canonize/src/enum.rs

```diff
@@ -0,0 +1,184 @@
+            type Output = #humanized;
+
+            fn humanize(self, api: &impl cosmwasm_std::Api) -> cosmwasm_std::StdResult<Self::Output> {
+                Ok(match self {
+                    #humanized_impls
+                })
+            }
+        }
+    })
+}
+
+fn add_serde_derive(strukt: &mut ItemEnum) {
+    strukt.attrs.clear();
+    strukt.attrs.push(parse_quote!(#[derive(serde::Serialize)]));
+    strukt
+        .attrs
+        .push(parse_quote!(#[derive(serde::Deserialize)]));
+}
+
+fn transform_fields(fields: &mut Fields) -> syn::Result<()> {
+    match fields {
+        Fields::Named(fields) => {
+            for mut field in fields.named.iter_mut() {
+                let ty = &field.ty;
+                field.ty = parse_quote!(<#ty as fadroma::prelude::Canonize>::Output);
+            }
+        }
+        Fields::Unnamed(fields) => {
+            for mut field in fields.unnamed.iter_mut() {
+                let ty = &field.ty;
+                field.ty = parse_quote!(<#ty as fadroma::prelude::Canonize>::Output);
+            }
+        }
+        Fields::Unit => {
+            return Err(syn::Error::new(
+                Span::call_site(),
+                "Unit variants are not supported.",
+            ))
+        }
+    }
+
+    Ok(())
+}
+
+pub fn generate_match_arms(
+    variants: &mut Punctuated<Variant, Comma>,
+    from: &Ident,
+    to: &Ident,
+    canonize: bool,
+) -> proc_macro2::TokenStream {
+    let mut res = Vec::new();
+    for (_, variant) in variants.iter_mut().enumerate() {
+        let name = &variant.ident;
+
+        let names = extract_field_names(&variant.fields);
+
+        let fields = canonize_fields(&variant.fields, canonize);
+
+        res.push(quote! {
+            #from::#name #names => #to::#name #fields,
+        })
+    }
+
+    quote! {
+        #(#res)*
+    }
+}
+
+fn extract_field_names(fields: &Fields) -> proc_macro2::TokenStream {
+    match &fields {
+        Fields::Named(fields) => {
+            let mut members: Punctuated<FieldValue, Comma> = Punctuated::new();
+            for field in fields.named.iter() {
+                // Named field, unwrap is fine here.
+                let name = field.ident.as_ref().unwrap();
+                let value = parse_quote!(#name);
+
+                members.push(value);
+            }
+
+            quote!({ #members })
+        }
+        Fields::Unnamed(fields) => {
+            let mut members: Punctuated<Expr, Comma> = Punctuated::new();
+            for i in 0..fields.unnamed.len() {
+                let ident = Ident::new(&format!("{}{}", "var", &i.to_string()), Span::call_site());
+                let value = parse_quote!(#ident);
+
+                members.push(value);
+            }
+
+            quote!((#members))
+        }
+        Fields::Unit => {
+            unreachable!()
+        }
+    }
+}
+
+fn canonize_fields(fields: &Fields, canonize: bool) -> proc_macro2::TokenStream {
```

Comment:
This also looks the same as: https://github.com/hackbg/fadroma/blob/60c09b1a64a0341ea3d36d2e5e4f61f78b7ff889/crates/fadroma-derive-canonize/src/common.rs#L10-L47